fix: TestProject.RuntimeTests do not run in stand alone test runner builds and fail on consoles - v1.1.0

NoelStephensUnity opened the pull request:
This includes the adjustments required to get the TestProject.RuntimeTests building as a stand alone test runner build:
[MTT-2313](https://jira.unity3d.com/browse/MTT-2313)
This includes the adjustments required to get the TestProject.RuntimeTests tests passing when ran in a stand alone test runner build (which should pass on consoles but is still TBD)
[MTT-2376](https://jira.unity3d.com/browse/MTT-2376)

These updates are based off of [PR-1644](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1644)  but might require additional changes. (WIP)

## Testing and Documentation
* No tests have been added.
* No documentation changes or additions were necessary.

## Verification Console Tests
[The results of this test](https://unity-ci.cds.internal.unity3d.com/job/11432691/dependency-graph) is #1653 and #1655 combined in order to make sure one set of changes does not impact the other
![image](https://user-images.githubusercontent.com/73188597/152644358-d987c3ce-c4e2-4c21-b23e-bbca071bccf3.png)

## Cherry Pick Squash and Merge Information
commit 455a75607dd50a7cd6f2fb083ebc6efc9010d7d0
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 12:19:02 2022 -0600

    Revert "test"

    This reverts commit a2f6fcac53a6d2f6d3facbbb52d3d11e613a2545.

commit 1c2d493ef29cd5c571fdcef0f95cb447acaab46a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 12:17:02 2022 -0600

    style

    white space update

commit ce84abab24b463cc8e5ef8c32a54fca5bdab93cc
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 11:34:22 2022 -0600

    test update

    This test was failing on some of the console platforms.  Made adjustments to avoid timing issues.

commit 8ad9b42141f5f69f60c76d099d50e1337e80442b
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 07:53:49 2022 -0600

    fix

    This is a work around for the MTT-1703 issue with RPCs.

commit 85fded20f49f073ff024eacc83ce728ba5c69550
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 05:40:31 2022 -0600

    style

    white spaces.

commit 82606ea2bddcb4ac659b550abed01a0a8a22f4b5
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 22:31:35 2022 -0600

    test update

    Still trying to get rid of IL2CPP error

commit 5915c3e0247cebf8828c208761d972c31257977c
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 20:13:45 2022 -0600

    test fix

    Trying to figure out weird ILCPP serialization issue.

commit 091d82fffb86d0eebbc846dc15f4dc2aedeada56
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 18:01:13 2022 -0600

    test refactor

    Needed some MULTIPLAYER_TOOLS helpers that I accidentally excluded.

commit a2f6fcac53a6d2f6d3facbbb52d3d11e613a2545
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 12:03:10 2022 -0600

    test

    disabling all tests in Unity.Netcode.RuntimeTest

commit 53194735f1afc1e166c561d8cfe20e45be84264a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 12:02:28 2022 -0600

    revert

    reverting as this was a dumb approach.

commit 312a5f5609ec62026b739a55dbdcd2f5564d0f42
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 11:19:59 2022 -0600

    test

    Seeing if this fixes the weird issue with the standards failing.

commit 76af4d409da3d44dfbb3ddaaace50c6ed8fe42db
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 09:08:26 2022 -0600

    test

    turning off the Unity.Netcode.RuntimeTests on console platforms temporarily in order to run just the TestProject.RuntimeTests on consoles.

commit 67566d634701690fefaff94c3282ceb930a2c799
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 09:06:27 2022 -0600

    refactor

    Removing assembly references not being used.

commit 2480daf97b7ccecaf09e870d6875630549bbce26
Merge: bee18d362 096f614bb
Author: Noel Stephens <noel.stephens@unity3d.com>
Date:   Tue Feb 1 20:36:59 2022 -0600

    Merge branch 'develop' into fix/test-roject-runtimetests-standalone-testrunner-build

commit bee18d362a041da8387f0f607126f02f415ebd01
Merge: 2fa12fa73 e07450fdf
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:28:11 2022 -0600

    Merge branch 'develop' into fix/test-roject-runtimetests-standalone-testrunner-build

commit 2fa12fa7369b5331fccd681f15bd807694d104a3
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:05:18 2022 -0600

    fix

    This fixes the issues with running the NetworkVariableInitializationOnNetworkSpawnTest when it was run in a stand alone test runner build which should be console friendly at this point.

commit c47ea65fef6a33264987ca4cb729879d1a2d0959
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:04:03 2022 -0600

    fix

    Fixing issues with the NetworkPrefabPoolAdditive class that is used in several tests including the NetworkSceneManagerTests.  The issues fixed were related to object references not being around (i.e. NetworkManager) as well as came up with a better way to detect if the coroutine was running in order to determine if it needed to be stopped.

commit 1ee12972d1cbee664e1a6a1829ae448dca60e7bd
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:56:57 2022 -0600

    fix

    The following updates to NetworkSceneManagerDDOLTests in NetworkSceneManagerTests.cs get it passing in a stand alone test runner build which should make it console friendly.

commit 0b955fecc99d8393a65a41bc6e2208ee1932d3d1
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:44:06 2022 -0600

    fix

    This fixes the issues with running NestedNetworkManagerTests in a stand alone test runner build and should run on consoles.

commit 2f18881b68df161ab51f11c106eed63a66b29187
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:30:41 2022 -0600

    fix

    The following updates to the MessageOrdering tests should get them working on consoles.

commit a967264a503a8a8df79e98d71d79206ef373df02
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:04:25 2022 -0600

    update

    Modifying the TestProject.RuntimeTests assembly to be a runtime only assembly.
    Had to migrate the TestProject.RuntimeTests out of the #if UNITY_EDITOR region in order to be able to compile a stand alone test runner build.

commit d8e5b1f8c1874d4015c66ba2d6820593f15344d2
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 12:22:06 2022 -0600

    refactor

    This commit changes the NetworkObjectParentingTests and ReparentingCubeNetBhv namespace to TestProject.RuntimeTests and uses the standard SceneManager to load the scene as opposed to the editor version.

commit d5efbe63718b9ce7994e8c42f5bf95222a7c5b2a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 12:14:48 2022 -0600

    refactor

    This commit migrates the NetworkPrefabGlobalObjectIdHashTests and NetworkObjectGlobalObjectIdHashTests into the TestProject.EditorTests assembly since they depend upon editor specific library calls and could be run as EditMode tests.

NoelStephensUnity committed:
Squashed commit of the following:

commit 455a75607dd50a7cd6f2fb083ebc6efc9010d7d0
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 12:19:02 2022 -0600

    Revert "test"

    This reverts commit a2f6fcac53a6d2f6d3facbbb52d3d11e613a2545.

commit 1c2d493ef29cd5c571fdcef0f95cb447acaab46a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 12:17:02 2022 -0600

    style

    white space update

commit ce84abab24b463cc8e5ef8c32a54fca5bdab93cc
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 11:34:22 2022 -0600

    test update

    This test was failing on some of the console platforms.  Made adjustments to avoid timing issues.

commit 8ad9b42141f5f69f60c76d099d50e1337e80442b
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 07:53:49 2022 -0600

    fix

    This is a work around for the MTT-1703 issue with RPCs.

commit 85fded20f49f073ff024eacc83ce728ba5c69550
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Thu Feb 3 05:40:31 2022 -0600

    style

    white spaces.

commit 82606ea2bddcb4ac659b550abed01a0a8a22f4b5
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 22:31:35 2022 -0600

    test update

    Still trying to get rid of IL2CPP error

commit 5915c3e0247cebf8828c208761d972c31257977c
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 20:13:45 2022 -0600

    test fix

    Trying to figure out weird ILCPP serialization issue.

commit 091d82fffb86d0eebbc846dc15f4dc2aedeada56
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 18:01:13 2022 -0600

    test refactor

    Needed some MULTIPLAYER_TOOLS helpers that I accidentally excluded.

commit a2f6fcac53a6d2f6d3facbbb52d3d11e613a2545
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 12:03:10 2022 -0600

    test

    disabling all tests in Unity.Netcode.RuntimeTest

commit 53194735f1afc1e166c561d8cfe20e45be84264a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 12:02:28 2022 -0600

    revert

    reverting as this was a dumb approach.

commit 312a5f5609ec62026b739a55dbdcd2f5564d0f42
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 11:19:59 2022 -0600

    test

    Seeing if this fixes the weird issue with the standards failing.

commit 76af4d409da3d44dfbb3ddaaace50c6ed8fe42db
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 09:08:26 2022 -0600

    test

    turning off the Unity.Netcode.RuntimeTests on console platforms temporarily in order to run just the TestProject.RuntimeTests on consoles.

commit 67566d634701690fefaff94c3282ceb930a2c799
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Wed Feb 2 09:06:27 2022 -0600

    refactor

    Removing assembly references not being used.

commit 2480daf97b7ccecaf09e870d6875630549bbce26
Merge: bee18d362 096f614bb
Author: Noel Stephens <noel.stephens@unity3d.com>
Date:   Tue Feb 1 20:36:59 2022 -0600

    Merge branch 'develop' into fix/test-roject-runtimetests-standalone-testrunner-build

commit bee18d362a041da8387f0f607126f02f415ebd01
Merge: 2fa12fa73 e07450fdf
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:28:11 2022 -0600

    Merge branch 'develop' into fix/test-roject-runtimetests-standalone-testrunner-build

commit 2fa12fa7369b5331fccd681f15bd807694d104a3
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:05:18 2022 -0600

    fix

    This fixes the issues with running the NetworkVariableInitializationOnNetworkSpawnTest when it was run in a stand alone test runner build which should be console friendly at this point.

commit c47ea65fef6a33264987ca4cb729879d1a2d0959
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 19:04:03 2022 -0600

    fix

    Fixing issues with the NetworkPrefabPoolAdditive class that is used in several tests including the NetworkSceneManagerTests.  The issues fixed were related to object references not being around (i.e. NetworkManager) as well as came up with a better way to detect if the coroutine was running in order to determine if it needed to be stopped.

commit 1ee12972d1cbee664e1a6a1829ae448dca60e7bd
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:56:57 2022 -0600

    fix

    The following updates to NetworkSceneManagerDDOLTests in NetworkSceneManagerTests.cs get it passing in a stand alone test runner build which should make it console friendly.

commit 0b955fecc99d8393a65a41bc6e2208ee1932d3d1
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:44:06 2022 -0600

    fix

    This fixes the issues with running NestedNetworkManagerTests in a stand alone test runner build and should run on consoles.

commit 2f18881b68df161ab51f11c106eed63a66b29187
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:30:41 2022 -0600

    fix

    The following updates to the MessageOrdering tests should get them working on consoles.

commit a967264a503a8a8df79e98d71d79206ef373df02
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 17:04:25 2022 -0600

    update

    Modifying the TestProject.RuntimeTests assembly to be a runtime only assembly.
    Had to migrate the TestProject.RuntimeTests out of the #if UNITY_EDITOR region in order to be able to compile a stand alone test runner build.

commit d8e5b1f8c1874d4015c66ba2d6820593f15344d2
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 12:22:06 2022 -0600

    refactor

    This commit changes the NetworkObjectParentingTests and ReparentingCubeNetBhv namespace to TestProject.RuntimeTests and uses the standard SceneManager to load the scene as opposed to the editor version.

commit d5efbe63718b9ce7994e8c42f5bf95222a7c5b2a
Author: NoelStephensUnity <73188597+NoelStephensUnity@users.noreply.github.com>
Date:   Tue Feb 1 12:14:48 2022 -0600

    refactor

    This commit migrates the NetworkPrefabGlobalObjectIdHashTests and NetworkObjectGlobalObjectIdHashTests into the TestProject.EditorTests assembly since they depend upon editor specific library calls and could be run as EditMode tests.

NoelStephensUnity committed:
fix

NetworkVariableInitOnNetworkSpawn needed to be updated.

NoelStephensUnity committed:
fix

disabling tests that were specific to snapshot.

NoelStephensUnity committed:
refactor

Migrating the DontShowInTransportDropdownAttribute into the SIPTransport in preparation of migrating all test helper related classes into their own assembly.

NoelStephensUnity committed:
refactor-2

missed an asmdef file.

NoelStephensUnity committed:
refactor

Realized I needed to create the separate assembly where test related helper classes exist in order to avoid issues with UTP tests and circular assembly definitions.

NoelStephensUnity committed:
Merge branch 'release/1.1.0' into fix/testproject-runtimetests-not-passing-on-all-console-platforms-v1.1.0

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/TestHelpers/com.unity.netcode.testhelpers.asmdef:
```diff
@@ -0,0 +1,14 @@
+{
```
This assembly was going to be created later, but I ended up having to create the very beginning of this assembly to avoid issues with UTP testing assemblies and circular reference issues.  All adjustments to com.unity.netcode.adapter.utp.<name> assemblies was to account for this update.

NoelStephensUnity commented on com.unity.netcode.adapter.utp/Editor/com.unity.netcode.adapter.utp.editor.asmdef:
```diff
@@ -1,22 +1,23 @@
-  "defineConstraints": [],
-  "versionDefines": [],
-  "noEngineReferences": false
+    "name": "Unity.Netcode.Adapter.UTP.Editor",
```
See: [com.unity.netcode.gameobjects/Tests/TestHelpers/com.unity.netcode.testhelpers.asmdef](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1655/files#diff-0b47e9df2c832a0d09b302203e0e37b07aa10271c3db562a76da031cd20922c7)

NoelStephensUnity commented on com.unity.netcode.adapter.utp/Tests/Editor/com.unity.netcode.adapter.utp.editortests.asmdef:
```diff
@@ -4,12 +4,24 @@
     "references": [
```
See: [com.unity.netcode.gameobjects/Tests/TestHelpers/com.unity.netcode.testhelpers.asmdef](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1655/files#diff-0b47e9df2c832a0d09b302203e0e37b07aa10271c3db562a76da031cd20922c7)

simon-lemay-unity reviewed (approved):
Looks good to me for the adapter changes. Haven't reviewed the rest.

NoelStephensUnity committed:
fix

Make the Unity.Netcode.TestHelpers:
-visible to Unity.Netcode.Runtime.
-a testing assembly
-reference the same assemblies that the Unity.Netcode.RuntimeTests assembly does.
-include the same versionDefines for the multiplayer tools define.

Noticed duplicate references and defined contraints in the netcode editortests assembly, so removed those.

NoelStephensUnity committed:
fix

Had to create a place holder class to assure the Unity.Netcode.TestHelpers existed in stand alone test runner builds.

NoelStephensUnity committed:
test-fix

At this point I am just sort of bumping through figuring out where the dependency issue is failing.

NoelStephensUnity committed:
update

missing files and asmdef updated

NoelStephensUnity committed:
update

Still trying to figure out where the disconnect is with the  com.unity.netcode.adapter.utp package tests and the Unity.Netcode.TestHelpers assembly...it seems to be the failing

NoelStephensUnity committed:
update

Seeing if this is an issue with things being compiled away/excluded or the like.

NoelStephensUnity committed:
Update PlaceHolderForNameSpace.cs

NoelStephensUnity committed:
update

Maybe changing the UTP editor namespace?

NoelStephensUnity committed:
fix

Fixing some internals issues.

NoelStephensUnity committed:
update

removing references to TestHelpers from UTP editor and runtime assemblies.

NoelStephensUnity committed:
update

one more...*sigh*

NoelStephensUnity committed:
update

NoelStephensUnity committed:
update

Perhaps autoreference was not properly set and noticed overrideReferences was improperly set on some assemblies when it had no assemblies that were assigned in the assembly references.
